### ISSUE TITLE
Remove deprecated (and to be removed) std::allocator<void>

### ISSUE THREAD

**Issue by tobiasneumann** [ros2/rosidl#770]
## Bug report

**Required Info:**

- Operating System:
  - Debian 12
- Installation type:
  - source
- Version or commit hash:
  - HEAD
- DDS implementation:
  - N/A
- Client library (if applicable):
  - N/A

#### Steps to reproduce issue

Build this code with C++17 or higher and `-Wdeprecated-declarations`.

#### Expected behavior

It compiles without warnings.

#### Actual behavior

It compiles with warnings.

#### Additional information

`std::allocator<void>` is deprecated in C++17 and removed in C++20.
https://en.cppreference.com/w/cpp/memory/allocator

**Comment by clalancette** [comment]
Yeah, I see what you mean about the removal.  Do you happen to know what the replacement should be?  Also pinging @azeey , @wjwwood for advice.
